### PR TITLE
BGE: Remove "Update this Opportunity" link in BGE Donation Selector component when Opportunity has existing Payments

### LINKED ISSUE #4286: Batch Gift Entry: Update Opportunity Feature May Overwrite Opportunity Fields
If Payments are enabled in an org and a customer is using Batch Gift Entry, they have the option to Update Opportunity or Apply New Payment to a matched gift if the Opportunity has no payments or all payments have been paid. Selecting the “Update Opportunity” option may lead to unintended Opportunity field changes — Amount and Close Date, for instance. The Stage will also change to Closed Won.

Recent work will address this by changing the options that appear for a matched gift:

- When Payments are enabled and a matched Opportunity has no payments, we will only display an “Apply New Payment” option. We will remove the “Update Opportunity” link.

- When Payments are not enabled and a matched Opportunity has no payments, there will be an option to Update Opportunity.

### PR DESCRIPTION
**Pull request by kenlewis-sfdo**
- This update removes the "Update this Opportunity" link within the Review Donations component if the Opportunity has any related Payments.  Previously the "Update this Opportunity" link was exposed and could lead to an out-of-sync state between an Opportunity and its related paid Payments.

# Critical Changes
## Batch Gift Entry
In Batch Gift Entry, when users click Review Donations, they'll be able to update an Opportunity only if the Opportunity has no related Payments—paid or unpaid. Previously, a user could click Update this Opportunity for Opportunities with related Payments, and this could lead to an Opportunity's amount or stage out of sync with its related paid Payments. To check if you have any Opportunities out of sync with the related payments follow the steps in this Quip doc: https://salesforce.quip.com/0XoxAiKW2HMJ

# Changes

# Issues Closed

- Fixes #4286

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by melissabarber** [comment]
**lurch: attach W-033025

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E0000087uerQAA" target="_blank">W-033025</a>

File: src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp [hunk 1]
```diff
@@ -59,28 +59,28 @@
...
-                        <aura:if isTrue="{!equals(donation.Id, v.selectedDonation.Id)}">
+                        <aura:if isTrue="{!equals(payment.Id, v.selectedDonation.Id)}">
                             <div class="selectedDonationBody">
                                 <lightning:icon iconName="utility:check" size="small" class="slds-p-right_small selectedCheckmark"
                                                 alternativeText="{!$Label.c.stgCheckboxTrue}" title="{!$Label.c.stgCheckboxTrue}"/>
                                 <lightning:formattedText value="{!$Label.c.bgeFormSelectedPayment}"/>
                             </div>
                             <aura:set attribute="else">
                                 <lightning:button
-                                        value="{!donation}"
```
Comment (ptewson-sfdo) [inline src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp]: This renaming is a great idea for clarity!

File: src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp [hunk 2]
```diff
@@ -117,36 +117,38 @@
...
-                <lightning:card title="{!donation.Name}" iconName="standard:opportunity"
-                                class="{!'slds-card_boundary sameCardHeight' + if(equals(donation.Id, v.selectedDonation.Id), ' selectedDonationCard')}">
+                <lightning:card title="{!opportunityWrapper.opportunity.Name}" iconName="standard:opportunity"
+                                class="{!'slds-card_boundary sameCardHeight' + if(equals(opportunityWrapper.opportunity.Id, v.selectedDonation.Id), ' selectedDonationCard')}">
 
                     <aura:set attribute="footer">
                         <div class="selectedDonationBody">
                             <lightning:layout multipleRows="true">
+                                <aura:if isTrue="{!not(opportunityWrapper.hasPayments)}">
+                                    <lightning:layoutItem size="12" padding="around_xx-small"
```
Comment (ptewson-sfdo) [inline src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp]: Just curious, why did you bring the lightning:layoutItem inside the aura:if?

Comment (ptewson-sfdo) [reply src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp]: NM I think I get it - you are bringing an additional aura:if condition around "hasPayments", not moving the condition.

File: src/aura/BGE_EntryForm/BGE_EntryFormHelper.js [hunk 3]
```diff
@@ -90,20 +90,34 @@
...
-     * @description: queries open donations for upcoming donations
+     * @description: Gets open donations without any Payments and unpaid Payments
+     * for use in the Donation Selector component.
      * @return: void
      */
-    queryOpenDonations: function(component, donorId) {
+    getOpportunitiesWithOppPayments: function(component, recordId) {
         const donorType = component.get('v.donorType');
 
-        let action = component.get('c.getOpenDonations');
```
Comment (ptewson-sfdo) [inline src/aura/BGE_EntryForm/BGE_EntryFormHelper.js]: With this change, do we need to keep the getOpenDonations() method in the apex controller around? Should we clean that up or do we still need that?

Comment (ptewson-sfdo) [reply src/aura/BGE_EntryForm/BGE_EntryFormHelper.js]: And follow-up - it looks like there's several apex tests pointing to getOpenDonations, but we probably need similar coverage of the new method (and if we are responsibly removing that method, clearing out the dependent tests as well)

File: src/aura/BGE_EntryForm/BGE_EntryFormHelper.js [hunk 4]
```diff
@@ -90,20 +90,34 @@
...
             const state = response.getState();
             if (state === 'SUCCESS') {
-                const openDonations = JSON.parse(response.getReturnValue());
-                component.set('v.openOpportunities', openDonations.openOpportunities);
-                component.set('v.unpaidPayments', openDonations.unpaidPayments);
+                const opportunitiesWithPayments = JSON.parse(response.getReturnValue());
+                let openOpportunitiesWithoutAnyOpenPayments = [];
+                let unpaidPayments = [];
+                opportunitiesWithPayments.forEach(function (opportunityWrapper) {
+                        if (opportunityWrapper.hasPayments === false ||
```
Comment (psadie) [inline src/aura/BGE_EntryForm/BGE_EntryFormHelper.js]: Quick question around whether or not we put symbols in the next break when lines are getting long. Do we have a separate javascript style guide like the[ apex style guide](https://salesforce.quip.com/TkjIACNQIURb)? Did a quick search in quip and didn't come up with anything. Do we potentially have that somewhere else?

@ptewson-sfdo @kenlewis-sfdo

File: src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp (hunk 1 above)
Comment (kenlewis-sfdo) [reply src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp]: 👍🏽

File: src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp (hunk 2 above)
Comment (kenlewis-sfdo) [reply src/aura/BGE_DonationSelector/BGE_DonationSelector.cmp]: Yep, exactly - the layout should be unchanged.

File: src/aura/BGE_EntryForm/BGE_EntryFormHelper.js (hunk 3 above)
Comment (kenlewis-sfdo) [reply src/aura/BGE_EntryForm/BGE_EntryFormHelper.js]: Nope, the new `getOpportunitiesWithOppPayments` method completely replaces it.  I just pushed a round of commits that adds new test methods to cover the new method, and removes the old `queryOpenDonations` method and related test methods.

File: src/aura/BGE_EntryForm/BGE_EntryFormHelper.js (hunk 4 above)
Comment (ptewson-sfdo) [reply src/aura/BGE_EntryForm/BGE_EntryFormHelper.js]: @psadie great question - I think we we use the Google guidance on java style, but I might be conflating with a past project. That said, that's a great starting point: http://google.github.io/styleguide/jsguide.html

From that resource, "When a line is broken at an operator the break comes after the symbol" (so like Ken has it here). Which conveniently aligns with my own inclination, although I've heard arguments for putting boolean operators on the start of the newline for long conditionals in if statements that made some sense.

File: src/classes/BGE_DataImportBatchEntry_CTRL.cls [hunk 5]
```diff
@@ -127,62 +127,108 @@ public with sharing class BGE_DataImportBatchEntry_CTRL {
...
+
         String relationshipField;
-        if (donorType == 'Account1') {
+        if (token == Account.SObjectType) {
             relationshipField = 'AccountId';
-        } else if (donorType == 'Contact1') {
+        } else if (token == Contact.SObjectType) {
             relationshipField = 'Primary_Contact__c';
+        } else {
+            throw new AuraHandledException('SObjectType ' + token + ' is not supported.');
```
Comment (psadie) [inline src/classes/BGE_DataImportBatchEntry_CTRL.cls]: This else is super unlikely to happen and so we don't really need a custom label, right? I sifted through our custom labels and didn't find anything more generic around _Something something_ + ' is not supported.' Should we?

**Pull request merged**